crates/nu-protocol/src/engine/command.rs

```diff
@@ -71,6 +80,21 @@ pub trait Command: Send + Sync + CommandClone {
     fn search_terms(&self) -> Vec<&str> {
         vec![]
     }
+
+    fn command_type(&self) -> CommandType {
+        match (
+            self.is_builtin(),
+            self.is_custom_command(),
+            self.is_parser_keyword(),
+            self.is_plugin().is_some(),
+        ) {
+            (true, false, false, false) => CommandType::Builtin,
+            (false, true, false, false) => CommandType::Custom,
+            (_, false, true, false) => CommandType::Keyword,
+            (false, false, false, true) => CommandType::Plugin,
+            _ => CommandType::Other,
```

Comment:
Sorry the code in this PR is a bit rushed, hence Draft status. I discovered that Keyword is also a Builtin. If you like this direction, maybe someone who knows the language internals better could take a look at this match statement.